### PR TITLE
[AMDGPU] Use correct DWord for v_dot4 S0 operand 

### LINKED ISSUE #112941: [AMDGPU] Numerical error related to load/store vectorization
The following input example: https://gist.github.com/qedawkins/c620832f96a5c504295f9694cc8956e2

Which is approximately compiled from
```
// int8_t a[144]; // [2 x 8 x 3 x 3]
// int8_t b[144]; // [2 x 8 x 3 x 3]
// int c[2];
__global__ void helloworld(int8_t* a, int8_t* b, int* c)
{
  int id = hipThreadIdx_x
  if (id < 2) {
    for (int c = 0; c < 8; ++c) {
      for (int h = 0; h < 3; ++h) {
        for (int w = 0; w < 3; ++w) {
          int x = a[id * 72 + c * 9 + h * 3 + w];
          int y = b[id * 72 + c * 9 + h * 3 + w];
          c[id] += x * y;
        }
      }
    }
  }
}
```

When compiled with
```
llc -O3 module_test_dispatch_0_rocm_hsaco_fb.optimized.ll -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100
```
Produces incorrect numerics. For inputs
```
a = np.ones((2, 8, 3, 3), dtype=np.int8)
b = np.broadcast_to(np.array([1, 2, 1], dtype=np.int8), (2, 8, 3, 3))
```
In other words only the values of `b` vary along the inner most loop, this gives `c = [88, 88]` when the correct values should be `(1 + 2 + 1) * 3 * 8 = [96, 96]`

Compiling with
```
llc -O3 module_test_dispatch_0_rocm_hsaco_fb.optimized.ll -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100 -amdgpu-load-store-vectorizer=false
```
(turning off the load/store vectorizer) or just with `-O1` instead of `-O3` gives the correct values. I am not sure exactly how the generated code is miscompiling and don't know whether load/store vectorization is the root cause or if disabling it is just masking the problem. Further information about the downstream issue here: https://github.com/iree-org/iree/issues/18798

cc @MaheshRavishankar

### PR DESCRIPTION
**Pull request by jrbyrnes**
Fixes a copy-paste typo.

The typo resulted in producing bad v_perm based operands for the v_dot4 combine. When adding a corresponding byte pair to the v_dot byte pair chains, we must take note of the byte position in the corresponding source nodes. These byte positions are used to ensure we extract the correct DWord from the ultimate source, and formulate a correct perm_mask from the extracted DWord.

With the typo, we the S0 byte would used the DWord offset for the corresponding S1 byte. If this offset was not the same as the true DWord offset for the S0 byte, we would extract and use the wrong byte for S0 in the v_dot. 

Fixes https://github.com/llvm/llvm-project/issues/112941

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/idot4s.ll [hunk 1]
```diff
@@ -3450,4 +3450,850 @@ entry:
 }
 
 
+define amdgpu_kernel void @ByteOffsetCorrectness(ptr addrspace(1) noalias readonly align 16 %inptr0, ptr addrspace(1) noalias readonly align 16 %inptr1, ptr addrspace(1) noalias align 16 %inptr2) local_unnamed_addr {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/idot4s.ll:3453]: Comment the test. Can this be shrunk?

File: llvm/test/CodeGen/AMDGPU/idot4s.ll [hunk 2]
```diff
@@ -3450,4 +3450,850 @@ entry:
...
+  %l90 = load i8, ptr addrspace(1) %gep90, align 1
+  %op91 = sext i8 %l91 to i32
+  %op90 = sext i8 %l90 to i32
+  %mul9 = mul nsw i32 %op90, %op91
+  %ivadd9 = add i32 %mul9, %ivadd8
+  %limit = add nuw nsw i64 %phi1, 1
+  %exitcond.not = icmp eq i64 %limit, 8
+  br i1 %exitcond.not, label %.110, label %.preheader2
+
+.110:                                              ; preds = %.preheader2
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/idot4s.ll:4286]: Shouldn't need control flow

File: llvm/test/CodeGen/AMDGPU/idot4s.ll (hunk 1 above)
Comment (jrbyrnes) [reply llvm/test/CodeGen/AMDGPU/idot4s.ll]: The behavior is actually a bit difficult to reproduce with a small concise example -- optimizations / DAG building consistently transform the code in a way which hides the issue.

File: llvm/test/CodeGen/AMDGPU/idot4-combine.ll [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  %workitemx = tail call i32 @llvm.amdgcn.workitem.id.x()
+  %sworkitemx = sext i32 %workitemx to i64
+  %workitemy = tail call i32 @llvm.amdgcn.workitem.id.y()
+  %sworkitemy = sext i32 %workitemy to i64
+  %workitemz = tail call i32 @llvm.amdgcn.workitem.id.z()
+  %sworkitemz = sext i32 %workitemz to i64
+  %ivtemp0 = add i64 %sworkitemy, %sworkitemz
+  %ivtemp1 = shl nsw i64 %ivtemp0, 5
+  %iv = add nsw i64 %ivtemp1, %sworkitemx
+  %0 = mul nsw i64 %ivtemp0, 2304
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/idot4-combine.ll:68]: Use named values in tests

File: llvm/test/CodeGen/AMDGPU/idot4-combine.ll [hunk 4]
```diff
@@ -0,0 +1,116 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 --start-before=amdgpu-isel -verify-machineinstrs < %s | FileCheck -check-prefixes=GFX11 %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/idot4-combine.ll:2]:
```suggestion
; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 --start-before=amdgpu-isel < %s | FileCheck -check-prefixes=GFX11 %s
```

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/idot4-combine.ll:2]: Why do you need -start-before? Should avoid that

File: llvm/test/CodeGen/AMDGPU/idot4s.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/idot4s.ll]: But what transformations? You can usually massage the test to defeat them. e.g. use volatile loads.

If you avoid using the work item ID intrinsics, you can avoid more transforms

**Review by arsenm — APPROVED** [review]

**Comment by arsenm** [comment]
Should add the fixes issue number to the description

**Comment by arsenm** [comment]
Description should also be more clear this is a bug fix

**Pull request merged**